src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -306,6 +306,12 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
 
             foreach (LibraryDependency d in item.Data.Dependencies)
             {
+                // Central transitive dependencies should be considered only for root nodes
+                if (!isRoot && d.ReferenceType == LibraryDependencyReferenceType.None)
```

Comment:
I'm not sure the isRoot check is correct here. 

This might disable warnings for scenarios like  

Project -> A 1.0.0 -> B 1.0.0
                             -> C 1.0.0 -> B 2.0.0
                             
This should select B 1.0.0 and give you a downgrade warning.

Reply:
@nkolev92 Is in your example B 1.0.0, a central transitive dependency, i.e. is it defined in `Directory.Packages.props` but there is no explicit PackageReference in A?

Reply:
I think it is exactly opposite, `LibraryDependencyReferenceType.None`  is used exclusively for CPM and transitive dependency pinning: https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs#L342-L354 and https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L486-L496 and https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L486-L496

So I sustain that this condition is right and doesn't affect non-transitive pinning scenarios.